psl/parser-database/src/files.rs

```diff
@@ -1,27 +1,59 @@
 use crate::FileId;
+use diagnostics::Diagnostics;
 use schema_ast::ast;
 use std::ops::Index;
 
 /// The content is a list of (file path, file source text, file AST).
 ///
 /// The file path can be anything, the PSL implementation will only use it to display the file name
 /// in errors. For example, files can come from nested directories.
-pub(crate) struct Files(pub(super) Vec<(String, schema_ast::SourceFile, ast::SchemaAst)>);
+pub struct Files(pub Vec<(String, schema_ast::SourceFile, ast::SchemaAst)>);
 
 impl Files {
-    pub(crate) fn iter(&self) -> impl Iterator<Item = (FileId, &String, &schema_ast::SourceFile, &ast::SchemaAst)> {
+    /// Create a new Files instance from multiple files.
+    pub fn new(files: Vec<(String, schema_ast::SourceFile)>, diagnostics: &mut Diagnostics) -> Self {
+        let asts = files
+            .into_iter()
+            .enumerate()
+            .map(|(file_idx, (path, source))| {
+                let id = FileId(file_idx as u32);
+                let ast = schema_ast::parse_schema(source.as_str(), diagnostics, id);
+                (path, source, ast)
+            })
+            .collect();
+        Self(asts)
+    }
+
+    /// Iterate all parsed files.
+    #[allow(clippy::should_implement_trait)]
```

Comment:
Why not implement trait instead?

Reply:
Because of

```
`impl Trait` in associated types is unstable
see issue #63063 <https://github.com/rust-lang/rust/issues/63063> for more information
add `#![feature(impl_trait_in_assoc_type)]` to the crate attributes to enable
```

which appears when trying to write e.g.

```rust
impl std::iter::IntoIterator for Files {
    type Item = (FileId, String, schema_ast::SourceFile, ast::SchemaAst);
    type IntoIter = impl Iterator<Item = (FileId, String, schema_ast::SourceFile, ast::SchemaAst)>;

    fn into_iter(self) -> Self::IntoIter {
        self.0
            .into_iter()
            .enumerate()
            .map(|(idx, (path, contents, ast))| (FileId(idx as u32), path, contents, ast))
    }
}
```

Also, `iter` and `into_iter` were previously defined like this by Tom, and I'd assume the reason he didn't implement the traits is the same one as mine. Do you have any explicit recommendation to work around this, or is it ok to leave the clippy warning suppressors as committed?

Reply:
Ah, i see. LGTM with no comments then